stats/bridge_stats.go

```diff
@@ -0,0 +1,347 @@
+	// The Close and WriteControl methods can be called concurrently with all other methods,
+	// so the mutex is not used here
+	return w.conn.Close()
+}
+
+type errorMessage struct {
+	Listener string `json:"listener"`
+	Err      string `json:"error"`
+}
+
+type processedBlockMessage struct {
+	Listener       string `json:"listener"`
+	ProcessedBlock uint64 `json:"processedBlock"`
+}
+
+type BridgeInfo struct {
+	Node           string            `json:"node"`
+	Operator       string            `json:"operator"`
+	Version        string            `json:"version"`
+	LastError      map[string]string `json:"lastError"`
+	ProcessedBlock map[string]uint64 `json:"processedBlock"`
+	PendingTasks   int               `json:"pendingTasks"`
+	FailedTasks    int               `json:"failedTasks"`
+}
+
+type authMsg struct {
+	ID     string    `json:"id"`
+	Info   *NodeInfo `json:"info"`
+	Secret string    `json:"secret"`
+}
+
+type Service struct {
+	lock             sync.Mutex
+	chainId          string
+	node             string
+	operator         string
+	voter            string
+	version          string
+	pass             string
+	host             string
+	lastError        map[string]string
+	processedBlock   map[string]uint64
+	pendingTasks     uint64
+	failedTasks      uint64
+	pongCh           chan struct{}
+	errCh            chan errorMessage
+	processedBlockCh chan processedBlockMessage
+	quitCh           chan struct{}
+	store            stores.TaskStore
+}
+
+func NewService(node, chainId, operator, host, pass string, db *gorm.DB) {
+	BridgeStats = &Service{
+		node:             node,
+		chainId:          chainId,
+		operator:         operator,
+		pass:             pass,
+		host:             host,
+		store:            stores.NewTaskStore(db),
+		lastError:        make(map[string]string),
+		processedBlock:   make(map[string]uint64),
+		pongCh:           make(chan struct{}, 1),
+		errCh:            make(chan errorMessage, 1),
+		processedBlockCh: make(chan processedBlockMessage, 1),
+		quitCh:           make(chan struct{}, 1),
+	}
+}
+
+func (s *Service) Start() {
+	errTimer := time.NewTimer(0)
+	defer errTimer.Stop()
+	for {
+		select {
+		case <-s.quitCh:
+			return
+		case <-errTimer.C:
+			// Establish a websocket connection to the server on any supported URL
+			var (
+				conn *connWrapper
+				err  error
+			)
+			dialer := websocket.Dialer{HandshakeTimeout: 5 * time.Second}
+			header := make(http.Header)
+			header.Set("origin", "http://localhost")
+			c, _, e := dialer.Dial(s.host, header)
+			err = e
+			if err == nil {
+				conn = newConnectionWrapper(c)
+			}
+			if err != nil {
+				log.Warn("Stats server unreachable", "err", err)
+				errTimer.Reset(10 * time.Second)
+				continue
+			}
+
+			// Authenticate the client with the server
+			if err = s.login(conn); err != nil {
+				log.Warn("Stats login failed", "err", err)
+				conn.Close()
+				errTimer.Reset(10 * time.Second)
+				continue
+			}
+			go s.loop(conn)
+			s.readLoop(conn)
```

Comment:
There is a loop in readLoop function, so readLoop function does not likely to return. As a result, this cannot loop back to select and the quitCh closed status is not checked frequently.